packages/react/kibana_context/theme/theme_provider.tsx

```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import React, { useMemo } from 'react';
+import useObservable from 'react-use/lib/useObservable';
+
+import {
+  CurrentEuiBreakpointProvider,
+  EuiThemeProvider,
+  EuiThemeProviderProps,
+} from '@elastic/eui';
+
+import {
+  getColorMode,
+  defaultTheme,
+  type ThemeServiceStart,
+} from '@kbn/react-kibana-context-common';
+
+// Extract the `theme` from `EuiThemeProviderProps` as a type.
+type EuiTheme<T = {}> = EuiThemeProviderProps<T>['theme'];
+
+// Omit the `theme` and `colorMode` props from `EuiThemeProviderProps` so we can
+// add our own `euiTheme` prop and derive `colorMode` from the Kibana theme.
+interface EuiProps<T = {}> extends Omit<EuiThemeProviderProps<T>, 'theme' | 'colorMode'> {
+  euiTheme?: EuiTheme<T>;
+}
+
+/**
+ * Props for the `KibanaThemeProvider`.
+ */
+export interface KibanaThemeProviderProps extends EuiProps {
+  /** The `ThemeServiceStart` API. */
+  theme: ThemeServiceStart;
+}
+
+/**
+ * A Kibana-specific theme provider that uses the Kibana theme service to customize the EUI theme.
+ */
+export const KibanaThemeProvider = ({
```

Comment:
[attaching to random file/line for GH threading to address a question in the PR description]

> Does it make sense to expose a `useTheme` hook from `@kbn/react-kibana-context-theme` to replace `useEuiTheme`?

My obviously biased vote is no 😅 EUI gets plenty of Kibana support questions around theming/styling etc., and in general the fewer layers of abstraction that exist between EUI and Kibana devs, the easier it is for us to help/debug issues directly without having to know the specifics of Kibana internals.

That being said, if there's a reason for it, then sure - but I'd be curious to hear what that reason is and if EUI can help with our OOTB behavior rather than needing an extra abstraction.

Reply:
You're probably right.  My only thought on it was if there was any additional functionality around retrieving the theme-- or other metadata specific to Kibana-- that would benefit from the abstraction.  @kc13greiner actually _wrote_ a `useTheme` hook a few weeks ago.  We'll put the idea on ice until we have a solid use case.